instance/tests/models/test_openedx_ansible_mixins.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (C) 2015-2016 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+OpenEdXAppServer Ansible Mixin - Tests
+"""
+
+# Imports #####################################################################
+
+import os
+from unittest.mock import patch, call, Mock
+
+from instance.models.mixins.ansible import Playbook
+from instance.tests.base import TestCase
+from instance.tests.models.factories.openedx_appserver import make_test_appserver
+from instance.tests.utils import patch_services
+
+
+# Tests #######################################################################
+
+
+class AnsibleAppServerTestCase(TestCase):
+    """
+    Test cases for AnsibleAppServerMixin models
+    """
+    @patch_services
+    def test_inventory_str(self, mocks):
+        """
+        Ansible inventory string - should contain the public IP of the AppServer's VM
+        """
+        mocks.mock_create_server.side_effect = [Mock(id='test-inventory-server'), None]
+        mocks.os_server_manager.add_fixture('test-inventory-server', 'openstack/api_server_2_active.json')
+
+        appserver = make_test_appserver()
+        appserver.provision()  # This is when the server gets created
+        self.assertEqual(appserver.inventory_str, '[app]\n192.168.100.200')
+
+    @patch_services
+    def test_inventory_str_no_server(self, mocks):
+        """
+        Ansible inventory string - should raise an exception if the server has no public IP
+        """
+        appserver = make_test_appserver()
+        with self.assertRaises(RuntimeError) as context:
+            self.assertEqual(appserver.inventory_str, '[app]\n')
+        self.assertEqual(str(context.exception), "Cannot prepare to run playbooks when server has no public IP.")
+
+    @patch('instance.models.mixins.ansible.poll_streams')
+    @patch('instance.models.openedx_appserver.OpenEdXAppServer.inventory_str')
+    @patch('instance.models.mixins.ansible.ansible.run_playbook')
+    @patch('instance.models.mixins.ansible.open_repository')
+    def test_provisioning(self, mock_open_repo, mock_run_playbook, mock_inventory, mock_poll_streams):
+        """
+        Test instance provisioning
+        """
+        appserver = make_test_appserver()
+        mock_open_repo.return_value.__enter__.return_value.working_dir = '/cloned/configuration-repo/path'
```

Comment:
> I think it would make sense to introduce a variable for '/cloned/configuration-repo/path'

That wasn't my code but is an easy fix, so I'll do that.